core/src/test/scala/kafka/tools/TestLogCleaning.scala

```diff
@@ -0,0 +1,361 @@
+    //if you change this line, we need to update test_log_compaction_tool.py system test
+    println("Data verification is completed")
+  }
+
+  def createTopics(brokerUrl: String, topics: Seq[String]): Unit = {
+    val adminConfig = new Properties
+    adminConfig.put(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, brokerUrl)
+    val adminClient = admin.AdminClient.create(adminConfig)
+
+    val topicConfigs = Map(TopicConfig.CLEANUP_POLICY_CONFIG -> TopicConfig.CLEANUP_POLICY_COMPACT)
+    val newTopics = topics.map(name => new NewTopic(name, 1, 1).configs(topicConfigs.asJava)).asJava
+    adminClient.createTopics(newTopics).all.get()
+
+    TestUtils.waitUntilTrue(() => {
+        val newTopics = adminClient.listTopics.names.get()
+        topics.forall(topicName => newTopics.contains(topicName))
+    }, "timed out waiting for topics")
+
+    adminClient.close()
+  }
+
+  def dumpLog(dir: File) {
+    require(dir.exists, "Non-existent directory: " + dir.getAbsolutePath)
+    for (file <- dir.list.sorted; if file.endsWith(Log.LogFileSuffix)) {
+      val fileRecords = FileRecords.open(new File(dir, file))
+      for (entry <- fileRecords.records.asScala) {
+        val key = readString(entry.key)
+        val content =
+          if (!entry.hasValue)
+            null
+          else
+            readString(entry.value)
+        println("offset = %s, key = %s, content = %s".format(entry.offset, key, content))
+      }
+    }
+  }
+
+  def lineCount(file: File): Int = io.Source.fromFile(file).getLines.size
+
+  def validateOutput(producedDataFile: File, consumedDataFile: File) {
+    val producedReader = externalSort(producedDataFile)
+    val consumedReader = externalSort(consumedDataFile)
+    val produced = valuesIterator(producedReader)
+    val consumed = valuesIterator(consumedReader)
+
+    val producedDedupedFile = new File(producedDataFile.getAbsolutePath + ".deduped")
+    val produceWriter : OutputStreamWriter = new OutputStreamWriter(new FileOutputStream(producedDedupedFile), UTF_8)
+    val producedDeduped = new BufferedWriter(produceWriter, 1024 * 1024)
+
+    val consumedDedupedFile = new File(consumedDataFile.getAbsolutePath + ".deduped")
+    val consumeWriter : OutputStreamWriter = new OutputStreamWriter(new FileOutputStream(consumedDedupedFile), UTF_8)
+    val consumedDeduped = new BufferedWriter(consumeWriter, 1024 * 1024)
+    var total = 0
+    var mismatched = 0
+    while (produced.hasNext && consumed.hasNext) {
+      val p = produced.next()
+      producedDeduped.write(p.toString)
+      producedDeduped.newLine()
+      val c = consumed.next()
+      consumedDeduped.write(c.toString)
+      consumedDeduped.newLine()
+      if (p != c)
+        mismatched += 1
+      total += 1
+    }
+    producedDeduped.close()
+    consumedDeduped.close()
+    println("Validated " + total + " values, " + mismatched + " mismatches.")
+    require(!produced.hasNext, "Additional values produced not found in consumer log.")
```

Comment:
I don't think require is quite the right thing here, since that's similar to "assert" -- supposedly for debugging.  We probably should just have our own function that exits with a nonzero return code if the test fails.

Reply:
added a function to verify condition and exits with a nonzero return code if condition  fails.